mat64/io.go

```diff
@@ -5,14 +5,159 @@
+//
+// Dense is little-endian encoded as follows:
+//  [ 0 -  8)  number of rows    (int64)
+//  [ 8 - 16)  number of columns (int64)
+//  [16 - ..)  matrix data elements (float64)
+//             [0,0] [0,1] ... [0,ncols-1]
+//             [1,0] [1,1] ... [1,ncols-1]
+//             ...
+//             [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, m.mat.Rows*m.mat.Cols*sizeFloat64+2*sizeInt64))
+	err := binary.Write(buf, binary.LittleEndian, int64(m.mat.Rows))
+	if err != nil {
+		return nil, err
+	}
+	err = binary.Write(buf, binary.LittleEndian, int64(m.mat.Cols))
+	if err != nil {
+		return nil, err
+	}
+
+	nrows, ncols := m.Dims()
+	for i := 0; i < nrows; i++ {
+		for j := 0; j < ncols; j++ {
+			err = binary.Write(buf, binary.LittleEndian, m.at(i, j))
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+// It panics if the receiver is a non-zero Dense matrix.
+//
+// See MarshalBinary for the on-disk layout.
+//
+// Limited checks on the validity of the binary input are performed:
+//  - matrix.ErrShape is returned if the number of rows or columns is negative,
+//  - matrix.ErrShape is returned if the resulting Dense matrix is too
+//  big for the current architecture (e.g. a 16GB matrix written by a
+//  64b application and read back from a 32b application.)
+func (m *Dense) UnmarshalBinary(data []byte) error {
+	if !m.isZero() {
+		panic("mat64: unmarshal into non-zero matrix")
+	}
+
+	buf := bytes.NewReader(data)
+	var rows int64
+	err := binary.Read(buf, binary.LittleEndian, &rows)
+	if err != nil {
+		return err
+	}
+	var cols int64
+	err = binary.Read(buf, binary.LittleEndian, &cols)
+	if err != nil {
+		return err
+	}
+
+	m.mat.Rows = int(rows)
+	m.mat.Cols = int(cols)
+	m.mat.Stride = int(cols)
+	m.capRows = int(rows)
+	m.capCols = int(cols)
+
+	// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+	const maxLen = int64(int(^uint(0) >> 1))
+
+	size := rows * cols
+	if rows < 0 || cols < 0 || int(size) < 0 || size > maxLen {
+		return matrix.ErrShape
+	}
+
+	m.mat.Data = use(m.mat.Data, int(size))
+
+	err = binary.Read(buf, binary.LittleEndian, &m.mat.Data)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Vector is little-endian encoded as follows:
+//   [0 -  8)  number of elements     (int64)
+//   [8 - ..)  vector's data elements (float64)
+func (v Vector) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, sizeInt64+v.n*sizeFloat64))
+	err := binary.Write(buf, binary.LittleEndian, int64(v.n))
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < v.n; i++ {
+		err = binary.Write(buf, binary.LittleEndian, v.mat.Data[i*v.mat.Inc])
```

Comment:
We don't do internal bounds checks unless the bounds tag is set, and the
function call will be inlined.